src/cargo/core/compiler/context/compilation_files.rs

```diff
@@ -206,9 +210,7 @@ impl<'a, 'cfg: 'a> CompilationFiles<'a, 'cfg> {
         // we don't want to link it up.
         if out_dir.ends_with("deps") {
             // Don't lift up library dependencies
-            if self.ws.members().find(|&p| p == unit.pkg).is_none() && !unit.target.is_bin() {
-                None
-            } else {
+            if self.roots.contains(unit) {
```

Comment:
I think this is a *little* suspect in the sense that this is pretty carefully crafted and I'm wary to switch wholesale to `roots`. Could we perhaps start off by basically skipping any rlib units? That is, we only care about binaries in the main workspace, mostly (and dylib/staticlib/cdylib/etc)

Reply:
We can do that, though I felt like that would be a bigger change.  It would still leave behind the race condition for the other output types (particularly `rmeta` files, though they don't cause catastrophic failures).  

I made an experimental branch to try this here:  https://github.com/ehuss/cargo/commit/68e7d2e9233b842740a9fc09c18df715f7aaff93
It seems more risky to me, though.

It seems like a pretty heavy change just to avoid these non-panic dependencies.  I'm still trying to think if there's an easier way...

Reply:
Hm ok yeah it's definitely not worth this becoming a larger change. 

So today the logic for "should I hardlink this up" is basically "am I a workspace member or am I a binary". That encompasses far too much as workspace member rlibs/rmeta have no business being linked up.

So this is instead saying that "if you were originally requested you're linked up". That includes libraries and rmeta business but you can't request more than one (even though we may compile many), so we should be resolving the race this way.

Writing this all down, I *think* the only difference is that all other workspace members aren't automatically linked up. I don't think, however, that it was possible for a workspace member to generate a relevant binary/staticlib for us to link upwards.

Or well, in other words, I think that this *should* be correct. I can't think of a way in which it would be wrong at least! In light of that I think we should be good to merge as soon as beta is branched.

Reply:
I think what *might* go wrong are dynamic libraries. We add top target directory to `LD_LIBRARY_PATH`:

https://github.com/rust-lang/cargo/blob/6cd841fe4a55d21d8f9a74102e86acb9475d45da/src/cargo/core/compiler/compilation.rs#L154-L155

However, in the next line we add the `deps` dir as well, so we should be fine here as well!